Raising exceptions in test mode (Issue: #10079)

SahilC opened the pull request.

SahilC committed:
Adding check for TestMode

SahilC committed:
Added test mode parameter to package-manager

SahilC committed:
Added test mode to the atom-environment

50Wliu commented on src/atom-environment.coffee:
```diff
@@ -171,6 +171,7 @@ class AtomEnvironment extends Model
       devMode, @configDirPath, resourcePath, safeMode, @config, styleManager: @styles,
       commandRegistry: @commands, keymapManager: @keymaps, notificationManager: @notifications,
       grammarRegistry: @grammars, deserializerManager: @deserializers, viewRegistry: @views
+      testMode: @inTestMode()
```
To be consistent with the rest I would change this to be `@testMode` instead.

SahilC committed:
Updated @inTestMode() => @testMode

SahilC commented:
I've made the requested changes, could someone verify if everything is ok?

thomasjo commented:
Can you explain the purpose of this PR? How is this different to [spec mode](https://github.com/atom/atom/blob/8229924b96497f0d0d0db47355d3e821c58a77ec/src/atom-environment.coffee#L427-L429)?

thomasjo commented:
OK, I see from the title that this is meant to fix #10079. Next time, please add all relevant details to the description body of the PR. I'll add some review comments for the remainder of my review.

thomasjo commented on src/atom-environment.coffee:
```diff
@@ -171,6 +171,7 @@ class AtomEnvironment extends Model
       devMode, @configDirPath, resourcePath, safeMode, @config, styleManager: @styles,
       commandRegistry: @commands, keymapManager: @keymaps, notificationManager: @notifications,
       grammarRegistry: @grammars, deserializerManager: @deserializers, viewRegistry: @views
+      testMode: @testMode
```
This needs to be removed. We already have this in the form of `specMode`.

thomasjo commented on src/atom-environment.coffee:
```diff
@@ -420,6 +421,10 @@ class AtomEnvironment extends Model
+  # Public: Returns a {Boolean} that is `true` if the current window is in test mode.
+  inTestMode: ->
+    @testMode ?= @getLoadSettings().testMode
+
```
Same as earlier, needs to be removed.

thomasjo commented on src/package-manager.coffee:
```diff
@@ -30,7 +30,7 @@ module.exports =
   constructor: (params) ->
     {
-      configDirPath, @devMode, safeMode, @resourcePath, @config, @styleManager,
+      configDirPath, @devMode, safeMode, @testMode, @resourcePath, @config, @styleManager,
```
`@testMode` needs to be removed.

thomasjo commented on src/package-manager.coffee:
```diff
@@ -397,7 +397,7 @@ class PackageManager
       options = {
         path: packagePath, metadata, packageManager: this, @config, @styleManager,
-        @commandRegistry, @keymapManager, @devMode, @notificationManager,
+        @commandRegistry, @keymapManager, @devMode, @testMode, @notificationManager,
```
Bye bye, `@testMode`

thomasjo commented on src/package.coffee:
```diff
@@ -32,7 +32,7 @@ class Package
     {
       @path, @metadata, @packageManager, @config, @styleManager, @commandRegistry,
-      @keymapManager, @devMode, @notificationManager, @grammarRegistry, @themeManager,
+      @keymapManager, @devMode, @testMode, @notificationManager, @grammarRegistry, @themeManager,
```
You know the drill by now 😄

thomasjo commented on src/package.coffee:
```diff
@@ -124,7 +124,10 @@ class Package
               @activateNow()
           catch error
-            @handleError("Failed to activate the #{@name} package", error)
+            if @testMode
```
Change `@testMode` to `@specMode`.

thomasjo replied:
Sorry I was mistaken earlier, but `@testMode` can be replaced with `atom.inSpecMode()`.

50Wliu commented:
Ah hah, I thought there was a specMode already defined somewhere...but I didn't see anywhere in the diff :P.

SahilC commented:
Yeah, I was confused about that too! Should I just rename everything to specMode?

thomasjo commented:
@SahilC It already exists, so remove almost all additions (as per my review comments). Keep the change package.coffee, but rename the `@testMode` there to `@specMode`.

SahilC commented:
But specMode isn't being passed to package.coffee's constructor. Am I missing something?

thomasjo commented:
@SahilC Sorry I should have looked at the diffs more closely, but you can use [`atom.inSpecMode()`](https://atom.io/docs/api/v1.11.2/AtomEnvironment#instance-inSpecMode).

SahilC committed:
Changed @testMode => atom.specMode

SahilC committed:
Bugfix for typo in atom environment

SahilC commented:
I've removed all occurrences of testMode, and replaced the check with atom.inSpecMode(). Wouldn't it be more consistent to use a variable rather than actually calling that function directly?

thomasjo commented:
I don't think it's worth adding the necessary changes for that given that we only check if we're in spec mode on a single line. I'd argue we should rather remove `@devMode` when we get around to converting this to JS.

This looks good to me now!

SahilC commented:
Cool, thanks!

maxbrunsfeld commented on src/package.coffee:
```diff
@@ -124,7 +124,10 @@ class Package
           catch error
-            @handleError("Failed to activate the #{@name} package", error)
+            if atom.inSpecMode()
+              raise error
```
This should be `throw error`.

maxbrunsfeld commented:
Thanks for looking into this problem @SahilC! Just one comment 👆 .

joh1234 commented:
Great, thanks. https://goo.gl/JBljdR
